proposals/2024-01-29_native_histograms_text_format.md

```diff
@@ -0,0 +1,90 @@
+## Native Histograms Text Format
+
+* **Owners:**
+  * Chris Marchbanks [@csmarchbanks](https://github.com/csmarchbanks)
+
+* **Implementation Status:** Not implemented
+
+* **Related Issues and PRs:**
+  * [Native Histogram Support in client_python](https://github.com/prometheus/client_python/issues/918)
+  * [OpenMetrics Protobuf format PR](https://github.com/OpenObservability/OpenMetrics/pull/256)
+
+* **Other docs or links:**
+  * [Design Doc for choosing a proposal](https://docs.google.com/document/d/1qoHf24cKMpa1QHskIjgzyf3oFhIPvacyJj8Tbe6fIrY/edit#heading=h.5sybau7waq2q)
+
+> TL;DR: This design doc is proposing a format for exposing native histograms in the OpenMetrics text format.
+
+## Why
+
+Today it is only possible to export native histograms using the Protocol Buffers (protobuf) scrape format. Many users prefer the text format, and some client libraries, such as the Python client, want to avoid adding a dependency on protobuf. 
+
+During a [dev summit in 2022](https://docs.google.com/document/d/11LC3wJcVk00l8w5P3oLQ-m3Y37iom6INAMEu2ZAGIIE/edit#bookmark=id.c3e7ur6rn5d2) there was consensus we would continue to support the text format. Including native histograms as part of the text format shows commitment to that consensus.
+
+### Pitfalls of the current solution
+
+Prometheus client libraries such as Python do not want to require a dependency on protobuf in order to expose native histograms, and in some languages protobuf is painful to use. Gating native histograms only to clients/users willing to use protobuf hurts adoption of native histograms, therefore, we would like a way to represent a native histogram in the text based format.
+
+## Goals
+
+Goals and use cases for the solution as proposed in [How](#how):
+
+* Support native histograms in the text format
+* (Secondary) Encode/decode efficiency
+* (Secondary) Ease of implementation for client libraries
+* (Secondary) Human readibility of the format
+
+Note that the goals of efficiency and human readability are commonly at odds with each other.
+
+### Audience
+
+Client library maintainers, OpenMetrics, and Prometheus scrape maintainers.
+
+## Non-Goals
+
+* Requiring backwards compatability (OpenMetrics 2.0 would be ok), and especially forwards compatability (not required in the OpenMetrics spec).
+
+## How
+
+Extend the OpenMetrics text format to allow structured values instead of only float values. This structured value will be used to encode a structure with the same fields as is exposed using the [protobuf exposition format](https://github.com/OpenObservability/OpenMetrics/pull/256). Starting with an example and then breaking up the format:
+```
+# TYPE nativehistogram histogram
+nativehistogram {count:24,sum:100,schema:0,zero_threshold:0.001,zero_count:4,positive_span:[0:2,1:2],negative_span:[0:2,1:2],positive_delta:[2,1,-2,3],negative_delta:[2,1,-2,3]}
```

Comment:
I think it would be helpful to add these examples:

* Native histogram with labels.
* Native histogram that also has classic histogram buckets
* Native histogram with multiple Exemplars but without classic histogram buckets

Reply:
Good catch. Native histograms can have repeated exemplars, and exemplars MUST have a timestamp, so the exemplar syntax is not trivially the same as for other samples.

Reply:
Happy to add examples for the first two, multiple exemplars are trickier though. In the linked design doc the thinking was that adding multiple exemplars is out of scope for the initial work and should be a separate proposal. See https://docs.google.com/document/d/1qoHf24cKMpa1QHskIjgzyf3oFhIPvacyJj8Tbe6fIrY/edit?disco=AAABFtLHXX0. Do you think we need to figure out text format support for multiple exemplars first?

Reply:
I have pushed a change adding multiple exemplar support as a non-goal, but I can figure something out if we decide it is an important goal to have as part of this work.

Reply:
By now, we have added multiple exemplars to the protobuf spec. Which means it's now way more interesting to do the same for the upcoming text format. However, it is not strictly required, either. If you think it complicates things significantly, you could prefer an incremental approach.

Reply:
Yeah, I want to keep the discussion separate. I think there are valid use cases for multiple exemplars on non-native histogram series as well, and don't want to bring that discussion into this format which is mostly agreed upon. If they both end up in OpenMetrics 1.1 that would be great though!